File: 2024/01/day_01.jl
function read_data(fn::String)::Array{Int64,2}
    lines = readlines(fn)
    first_numbers = []
    second_numbers = []

    for line in lines
        numbers = split(line)
        push!(first_numbers, parse(Int64, numbers[1]))
        push!(second_numbers, parse(Int64, numbers[2]))
    end

    first_numbers_sorted = sort(first_numbers)
    second_numbers_sorted = sort(second_numbers)

    return hcat(first_numbers_sorted, second_numbers_sorted)
end

function run_algorithm1(data::Array{Int64,2})::Int64
    l1 = data[:, 1]
    l2 = data[:, 2]
    
    @assert length(l1) == length(l2) "Lengths of l1 and l2 are not equal"

    sum = 0
    
    for i in 1:length(l1)
        sum += abs(l1[i] - l2[i])
    end

    return sum
end

function run_algorithm2(data::Array{Int64,2})::Int64
    l1 = data[:,1]
    l2 = data[:,2]

    @assert length(l1) == length(l2) "Lengths of l1 and l2 are not equal"

    freq_map = Dict{Int64, Int64}()
    for num in l2
        if haskey(freq_map, num)
            freq_map[num] += 1
        else
            freq_map[num] = 1
        end
    end

    sum = 0

    for i in 1:length(l2)
        if haskey(freq_map, l1[i])
            sum = sum + l1[i] * freq_map[l1[i]]
        end
    end
    return sum
end

data = read_data("input.txt")
println(run_algorithm1(data))
println(run_algorithm2(data))

File: 2024/01/day_01.py
from typing import List, Dict

INPUT_FILE = "input.txt"

def read_input(fn: str) -> List[List[int]]:
    output:List[List[int]] = []
    with open(fn) as f:
        for line in f:
            parts = [int(x) for x in line.strip().split()]
            for i, j in enumerate(parts):
                if len(output) <= i:
                    output.append([j])
                else:
                    output[i].append(j)
    return output

def run_algorithm(data: List[List[int]]) -> int:
    l1 = data[0]
    l2 = data[1]
    l1.sort()
    l2.sort()

    assert len(l1) == len(l2)

    return sum([abs(x - y) for x, y in zip(l1, l2)])

def run_algorithm2(data: List[List[int]]) -> int:
    l1 = data[0]
    l1.sort()

    h2: Dict[int, int] = {}
    for x in data[1]:
        h2[x] = h2.get(x, 0) + 1

    return sum([x * h2[x] for x in l1 if x in h2])

def main():
    data = read_input(INPUT_FILE)
    output = run_algorithm(data)
    print(f"final output part 1: {output}")
    output2 = run_algorithm2(data)
    print(f"final output part 2: {output2}")

if __name__ == "__main__":
    main()
